### ISSUE TITLE
[Feature Request] Additional fields in the backend views for better UX

### ISSUE THREAD

**Issue by Aka-shi** [hypertrace/hypertrace#169]
## Use Case
As part of driving the adoption of Hypertrace across our org, we wanted to improve the slow query debugging experience for developers on Hypertrace. As of today, in the backend explorer few fields are missing as first level columns and have to be queried explicitly. Few examples, 
- On the `/backends?time=1h` url, it is not intuitive if a backend is a HTTPS service or a Database. It might be good to have what kind of backend(if it is a database then if it is MySQL/PostgreSQL/etc,.) it is visible on this page so that the developer can easily find the backend related to his service.
- Having db.method, db.table and trace_id as first class columns on the `backends/backend/<id>/traces?time=1w` tab would help in narrowing down slow queries at a table and operation level much faster. As of today, user has to expand the trace and check the tags for a trace or they have to query using the query bar. 
- There is a bug in the trace explorer for backends because of which `span_id` is visible as `trace_id`

## Proposal
- Trace_id and span_id are already present in pinot schemas, so the changes required for these to be first class columns would only be in UI/Query layer. 
- By looking at the code [here](https://github.com/hypertrace/hypertrace-ingester/blob/main/hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/BackendType.java) It seems like a limited set of backend_types are supported today. Would need your opinion on how to proceed to add support for a broader range of backend types(MySQL/PostgreSQL/etc,.). 
- db.method, db.table are present in two different multi value columns today. Querying multi value columns from the UI would lead to higher query latencies and degraded UX at times. Would it be a good idea to have these fields as columns in the pinot backendEntityView schema? This columns can be null/zero-value for http/service backends or if the spans to the db are not instrumented in the apps.

**Comment by kotharironak** [comment]
@Aka-shi `There is a bug in the trace explorer for backends because of which span_id is visible as trace_id` has been taken care of in the latest release, you will able to see both trace_id and span_id 
<img width="1660" alt="Screenshot 2021-02-01 at 5 20 42 PM" src="https://user-images.githubusercontent.com/53209990/106455237-ec37c900-64b1-11eb-9c4a-7bbf09ff1430.png">

**Comment by kotharironak** [comment]
@buchi-busireddy @aaron-steinfeld @jcchavezs @rish691  Pl. have your suggestion/view if you think differently on this here.

**Comment by aaron-steinfeld** [comment]
> On the /backends?time=1h url, it is not intuitive if a backend is a HTTPS service or a Database. It might be good to have what kind of backend(if it is a database then if it is MySQL/PostgreSQL/etc,.) it is visible on this page so that the developer can easily find the backend related to his service.

This should already be visible via the type column on the backend list, although each type does need a bit of manual addition to associate an icon. If there's certain ones missing, please let us know:
![image](https://user-images.githubusercontent.com/45047841/106494579-74e14580-646f-11eb-9104-9b5f49085912.png)


> Having db.method, db.table and trace_id as first class columns on the backends/backend/<id>/traces?time=1w tab would help in narrowing down slow queries at a table and operation level much faster. As of today, user has to expand the trace and check the tags for a trace or they have to query using the query bar.

Agree on the first two, although I thing trace id is rarely useful at the row level. It's available on drill down, or to filter by, but generally not needed when viewing the list - do you have a use case for it?

As to implementing this, things get a bit trickier. While I agree they should be first class columns, I'm not sure I agree (not sure I disagree either, but think it's more questionable) that they should be first class attributes - since they only apply to a subset of traces (those with a head span exit call to a DB) and could be confusing in the general case. If we can get away with querying this data without adding new attributes without a perf hit, that'd be ideal.

**Comment by jcchavezs** [comment]
@Aka-shi thanks for putting up the issue. I think you are totally right, definition of backend is kind of implicit and you get used to it once you use HT but definitively counter intuitive for new comers (it was for me at the beginning). What I would suggest is that we have a classification of the backends as something like e.g. `storage`, `brokers`, etc and that would let us have first class attributes in the spans.

**Comment by Aka-shi** [comment]
>This should already be visible via the type column on the backend list

@aaron-steinfeld It does, but mysql/postgresql are part of `JDBC` backend resolver and would be visible as a `jdbc` backend on the UI too. The point I raised was to support these as their own backend types. But I agree, this can be a trivial change to end user and won't really help in improving the experience. Just a nice to have kind of feature.

Apart from the sql backends, Hypertrace as of today doesn't support other message brokers apart from rabbitmq. I think supporting Kafka, amazon sns/sqs(these are widely used in our org) as backend types would help the user. 

>As to implementing this, things get a bit trickier. While I agree they should be first class columns, I'm not sure I agree (not sure I disagree either, but think it's more questionable) that they should be first class attributes 

I share the same opinion on this. In the current setup, `db.xxx` tags would be relevant for only storage backends and having them as first class attributes might not be the best idea. But as @jcchavezs proposed, if we are classifying backends into `storage`, `brokers`, `service`, etc then having these as first class attributes makes sense.

**Comment by kotharironak** [comment]
I agree with the first part of enhancing the backend support for `kafka` and `sqs`.

For part-2 where we would like to make some fields as first-class, as of now, I can think of following options.

**Option 1**: Have columns based on db types  (as discussed above by @jcchavezs )
This suggests utilizing the `type` information in laying out column views.
e.g for db type `kafka`, we can show topic name as an additional column, and for db type `sql` we can show additional columns <db name, table name, db method, query>

**Option 2**: Continue with common column names across all backends types, but the value will be field according to db type.
e.g we can add two new columns for all db types <destination, query>. However, the value of this field will be filled during the enrichment process according to db types.

So, if we consider `destination` as a new column, and using https://github.com/open-telemetry/opentelemetry-specification/, we can do as below.

**destination : db type : value**
**kafka** : topic_name (messaging.destination)
**sqs** : queue_name (messaging.destination)

**grpc** : the endpoint which was accessed
**HTTP**:  the endpoint which was accessed

**mysql/sql** : dbname.tablename (extract from db.operation / db.statement)
**cassandra** : keyspacename.tablename ( db.cassandra.keyspace,  db.cassandra.table)

**mongodb** : collection name (db.mongodb)
**redis** : database index as described here - https://redis.io/commands/select (db.redis.database_index)


I also think on the traces list page of the backend, the column `type` seems to be redundant information as a given backend would be of a specific type and all of its traces would belong to the same type. So, we can remove this.
<img width="1675" alt="Screenshot 2021-03-01 at 3 46 22 PM" src="https://user-images.githubusercontent.com/53209990/109485789-17f5a100-7aa8-11eb-8bbb-1913945c44b4.png">

**Comment by JBAhire** [comment]
I can see option 1 being easy to adapt but It will need change in schema or using additional fields as tags. So, I am also inclined towards option 2 because we can tweak it more easily and enhance the current view itself with few additional columns.

Backend name should be picked from [db.system](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#connection-level-attributes) or [messaging.system](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#messaging-attributes) and we actually don’t need which shows this as all traces here will be from that particular backend itself.

Generic columns:
1. Name (name of database): This should be equal to `db.name` from otel specs. This will give us database name (eg., `customers`, `main`). And if we have users have more than one database under same backend, it makes complete sense to have this. For messaging systems this can fallback to `messaging.system` for now. Or we can use `messaging.destination` here and we won’t need Destination column at all.

2. Type (Type of Backend): Initially there can be 2 types, Database and messaging systems and eventually as we start supporting `FaaS` that can be one more category. Let’s keep categorisation more general here as we can make it depend on generic otel specification out of the box.

3. Query: this can be `db.statement` but should fall back to `db.operation` if `db.statement` is not set, which will be the case with db like mongo. And for messaging system it can be `messaging.operation`.

4. [Optional] Destination (Table name/ messaging system queue name): This can be `db.sql.table`/ `db.mongodb.collection` as Ronak also mentioned above. For messaging system it can fallback to `messaging.destination`. The only problem here is this is `Required` attribute for messaging system but not necessarily required but recommended in DBs, so there’s a chance that this column might remain in case of Databases if instrumentation is not done properly.

Summarizing it below,

| Column name | Details                                                                                       | Attributes                                                     |                                               |
|-------------|-----------------------------------------------------------------------------------------------|----------------------------------------------------------------|-----------------------------------------------|
|             |                                                                                               | **Database**                                                       | **Messaging**                                     |
| Name        | Name of database/ messaging service name or destination                                       | `db.name`                                                      | `messaging.system` or `messaging.destination` |
| Type        | DB/Messaging/FaaS**                                                                           |                                                                |                                               |
| Query       | Query                                                                                         | `db. statement` (fallback to `db.operation`)                   | `messaging.operation`                         |
| Destination | Operation and necessary in case of messaging systems. we can skip if we incorparate in Name.  | Table name using `db.sql.table` / `db.mongodb.collection` etc. | `messaging.destination`                       |

**Comment by aaron-steinfeld** [comment]
> 
> Generic columns:
> 
> Name (name of database): This should be equal to db.name from otel specs. This will give us database name (eg., customers, main). And if we have users have more than one database under same backend, it makes complete sense to have this. For messaging systems this can fallback to messaging.system for now. Or we can use messaging.destination here and we won’t need Destination column at all.
> 
> Type (Type of Backend): Initially there can be 2 types, Database and messaging systems and eventually as we start supporting FaaS that can be one more category. Let’s keep categorisation more general here as we can make it depend on generic otel specification out of the box.
> 
> Query: This can be db.statement but should fall back to db.operation if db.statement is not set, which will be the case with db like mongo. And for messaging system it can be messaging.operation.
> 
> [Optional] Destination (Table name/ messaging system queue name): This can be db.sql.table/ db.mongodb.collection as @kotharironak also mentioned above. For messaging system it can fallback to messaging.destination. The only problem here is this is Required attribute for messaging system but not necessarily required but recommended in DBs, so there’s a chance that this column might remain in case of Databases if instrumentation is not done properly.

So I think I was confused here - the conversation seems to go back and forth between discussing columns in the backend entity view and the backend trace view without specifying.

On the entity view, we have name and type already - we can (and should) improve the resolution of these values based on OTEL specs, but I don't think there's a new col to show here - destination and query may have many values for a single backend/row.

On the trace (call) view, type is redundant as @kotharironak pointed out and should be removed - but things like destination and query would be very helpful. "name" here seems to indicate something other than the backend name, but is likely also redundant if we have a destination and query col available.

**Comment by kotharironak** [comment]
@aaron-steinfeld The discussion here is in the context of backend traces (trace/call view) for additional information.

**Comment by JBAhire** [comment]
On further discussion around important uses cases, we have figured out some first-class information for quicker analysis in backend trace view which will answer questions like
- What kind of operations take more time?
- Which destination (table) has more errors?
- Which query takes more time?

we are moving forward with these changes
- Remove `TYPE` column from backend Trace (call) view (redundant)
- Keep the `NAME` column as it is as of now. (currently, it’s SPAN_NAME and exists in most of the tools) 
- Hold on for query columns as it is important but has many variants e.g select * from a customer where customer_id = <1234>. so, the template of the query is more important than the actual query. So, can be available as an extended view for now.
- Add destination column as mentioned above.
- Add Operation column to show db/messaging system operations

```
Destination 
<for backend type> : <suggested value for destination>
http_backend : API endpoint
sql : db-name.table-name 
mongo: db-name.collection-name
kafka: consumer-group-name.topic-name
SQS : <account_number>.queue-name
```
```
Operation:
<for backend type> : <suggested value for operation>
http_bakend : GET / POST 
sql : INSERT / UPDATE / DELETE
mongo : find / update / etc 
kafka : send / receive / delete / create / get
SQS : sendMessage / receiveMessage / enQueue / deQueue / get
```

**Comment by JBAhire** [comment]
Closing this issue as this is now available as a part of [Hypertrace Release 0.1.8](https://github.com/hypertrace/hypertrace/releases/tag/0.1.8). @Aka-shi , feel free to add fallback conditions in enricher as per the requirements. 

<img width="2325" alt="Screenshot 2021-03-19 at 9 04 12 PM" src="https://user-images.githubusercontent.com/26570044/111807439-f21a2a00-88f8-11eb-8c3c-159d548023de.png">
